Add blogid to the tracking object

AnnaMag opened the pull request:
#### Changes proposed in this Pull Request:

* Add `blogid` to the tracking object on the client side for events like page view and toggles. This allows to join the events triggered from the `wp-admin`  with the data about the site we record in Calypso on this id;

* extra: 
    move to the end the debug line in the `recordEvent` function called on the analytics object to enable viewing of updated object fields in their entirety;

#### Testing instructions:
* building JP off of this branch by the method of choice, verify that track events that are called via 
the `recordEvent`  function contain a `blog_id` field.

<img width="834" alt="Screenshot 2019-05-08 at 15 47 13" src="https://user-images.githubusercontent.com/13561163/57380117-9fbe3d00-71a8-11e9-830d-d7465a9c1d45.png">

The data is passed as an attribute of the event property object. It gets recorded in a respective field via Tracks ETL (which means it will land in the blogid column of the respective data tables, and seen as such in the live view).

Most notably, this fixes the blogid in:
Fixes https://github.com/Automattic/jetpack/issues/12256


<!-- Add the following only if this is meant to be in changelog -->
#### Proposed changelog entry for your changes:

AnnaMag committed:
Add blogid to the tracking object.

This allows to tie the events triggered from the wp-admin with the data we record in Calypso via the blog_id field.

AnnaMag committed:
Add extra parameter to the definition of the f-n initializing tracks object.

Move the line debugging to the console at the end of the function setting up
the properties of the tracking object for it to display the complete set.

tyxla commented on _inc/client/lib/analytics/index.js:
```diff
@@ -55,12 +56,17 @@ function buildQuerystringNoPrefix( group, name ) {
 	},
 
+	blogId( blogid ) {
+		_blog = { blogid: blogid };
```
`{ blogid: blogid }` could be simplified to `{ blogid }`

tyxla commented on _inc/client/lib/analytics/index.js:
```diff
@@ -55,12 +56,17 @@ function buildQuerystringNoPrefix( group, name ) {
 const analytics = {
 
-	initialize: function( userId, username, superProps ) {
+	initialize: function( blogid, userId, username, superProps ) {
```
This is changing the function signature and moves all the parameters, do you think we could add the new parameter to the end, or at least move it right before `superProps`?

AnnaMag replied:
addressed below

tyxla commented on _inc/client/lib/analytics/index.js:
```diff
@@ -102,22 +108,30 @@ const analytics = {
 				return;
 			}
 
+			if ( _blog ) {
```
This seems to be doing the same we're doing for `superProps`. Do you think we could set the blog ID as a superProp? This way we won't have to alter the `analytics.initialize` method signature, and will use what we already have.

AnnaMag replied:
yes, I agree.

AnnaMag replied:
I'd prefer to keep it at the same level as `userId` for the sake of future analyses (and data logic).

AnnaMag replied:
I'm not fully convinced this would be convenient on the data analysis end, but since here it makes sense and I'm roughly guessing, taking up on your suggestion :)

tyxla commented on _inc/client/main.jsx:
```diff
@@ -92,6 +92,7 @@ class Main extends React.Component {
 		const tracksUser = this.props.tracksUserData;
 		if ( tracksUser ) {
 			analytics.initialize(
+				tracksUser.blogid,
```
I'm seeing at least [1 more location](https://github.com/Automattic/jetpack/blob/master/_inc/idc-notice.js#L21) where we initialize the tracks client, and we haven't updated that to reflect the new parameter order. This will break things.

AnnaMag replied:
true, my bad

AnnaMag replied:
Solved as per remarks above and moving blogid to superProps.

AnnaMag replied:
moved to supeProps

tyxla commented on _inc/lib/tracks/class.tracks-client.php:
```diff
@@ -182,8 +182,12 @@ static function get_connected_user_tracks_identity() {
 		if ( ! $user_data = Jetpack::get_connected_user_data() ) {
 			return false;
 		}
+		if ( ! $blog_id = Jetpack_Options::get_option( 'id' ) ) {
```
Do we really want to make the blog ID required? Wont this break for cases where the site is not connected currently? TBH I expected the blog ID to be optional here.

AnnaMag replied:
good point.

tyxla commented on _inc/client/lib/analytics/index.js:
```diff
@@ -102,22 +108,30 @@ const analytics = {
 				eventProperties = assign( eventProperties, superProperties );
 			}
 
+			debug( 'Record event "%s" called with props %s', eventName, JSON.stringify( eventProperties ) );
```
👍 It's great that we're moving this to the bottom, so we'll be able to debug all the props we're actually logging, and not just a limited set of them.

tyxla commented on _inc/client/lib/analytics/index.js:
```diff
@@ -55,12 +56,17 @@ function buildQuerystringNoPrefix( group, name ) {
 		analytics.identifyUser();
 	},
 
+	blogId( blogid ) {
```
This won't be necessary if we use `superProps` to provide the blog ID instead.

AnnaMag committed:
Moving blogid to superProp at analytics initialization to use current function definition.

- also, make blogid information optional.

AnnaMag commented:
@tyxla, thanks for review. I'd appreciate another test when you get a chance.

tyxla commented on _inc/client/main.jsx:
```diff
@@ -93,7 +93,8 @@ class Main extends React.Component {
 				tracksUser.userid,
-				tracksUser.username
+				tracksUser.username,
+        tracksUser.blogid,
```
There seem to be some indentation issues here - this needs tabs instead of spaces.

tyxla commented on _inc/lib/tracks/class.tracks-client.php:
```diff
@@ -184,6 +184,7 @@ static function get_connected_user_tracks_identity() {
 		}
 
 		return array(
+			'blogid' => Jetpack_Options::get_option( 'id' ),
```
Are you planning to introduce any checks for the blog ID here? Because if we don't have it yet, this will fall back to `false`  and not `0` - I'd expect we always want to keep it a numeric value.

AnnaMag replied:
as mentioned yesterday, yes, I was meaning to add a condition.

tyxla reviewed (dismissed):
There is a at least [1 more instance](https://github.com/Automattic/jetpack/blob/f8078c2cd12ac508334da2fb08e37a92cf283c14/_inc/idc-notice.js#L21) of `analytics.initialize()` that we haven't updated to setup the blog ID accordingly.

oskosk committed:
Fix whitespace

tyxla reviewed (commented):
Even when manually fixing the commas I've mentioned in the review, I'm unable to get it working, I encounter the following error:

![](https://cldup.com/dx7cbjZDPF.png)

Have you tried rebasing perhaps?

AnnaMag commented:
Removed commas and rebased, no console errors on my end prior to that and after the changes. 
Tesing using docker.

@oskosk, could you please have a look to possibly shed some more light on the failures? and test would be appreciated. I cannot reproduce the errors Marin sees.

sirreal commented:
5baea01 seems to fix the travis errors from php 5.2

sirreal reviewed (changes requested):
Should this be updated as well?

https://github.com/Automattic/jetpack/blob/c36c5e2847a517a04f291f72f941072a8dc8e343/modules/widgets/search/js/search-widget-admin.js#L14-L17

In search-widget-admin and idc-notice, it's not clear to me which analytics might be used since they're just JS globals 😞 (there are [two](https://github.com/Automattic/jetpack/blob/master/_inc/lib/tracks/tracks-callables.js) JS [files](https://github.com/Automattic/jetpack/blob/master/_inc/client/lib/analytics/index.js) that export analytics)

It appears that the problem is that the `initialize` method has not been updated to accept the blog ID. It expects `superProps` in the third position, which sets a module variable. It then attempts to call `getAll` on this variable (a number, i.e. `0.getAll()`), which results in an error.

https://github.com/Automattic/jetpack/blob/c36c5e2847a517a04f291f72f941072a8dc8e343/_inc/client/lib/analytics/index.js#L58-L62

In some quick investigation, `superProps` doesn't appear to be used, but don't take my work for it. We need to take care if we're changing the `initialize` method signature to ensure the new signature works well, but we don't break existing calls.

sirreal commented on _inc/idc-notice.js:
```diff
@@ -18,7 +18,11 @@
 	// Initialize Tracks and bump stats.
 	if ( 'undefined' !== typeof analytics ) {
-		analytics.initialize( tracksUser.userid, tracksUser.username );
+		analytics.initialize(
```
Some searching suggests that this and the `seach-widget-admin` I mentioned in my review actually appear to use [tracks-callables](https://github.com/Automattic/jetpack/blob/master/_inc/lib/tracks/tracks-callables.js), which isn't updated at all for `blogid`.

https://github.com/Automattic/jetpack/blob/c36c5e2847a517a04f291f72f941072a8dc8e343/class.jetpack-idc.php#L305
https://github.com/Automattic/jetpack/blob/c36c5e2847a517a04f291f72f941072a8dc8e343/modules/widgets/search.php#L107

AnnaMag replied:
great catch @sirreal, we missed that. I'm working on updating the patch.

sirreal replied:
We might investigate why `superProps` was added and whether it's used. I have no idea how it might be used. `getAll` doesn't clarify exactly what is expected to be passed there.

sirreal commented on _inc/lib/tracks/class.tracks-client.php:
```diff
@@ -184,6 +184,9 @@ static function get_connected_user_tracks_identity() {
 
 		return array(
+			'blogid' => ! empty( Jetpack_Options::get_option( 'id' ) )
+				? Jetpack_Options::get_option( 'id' )
```
This is causing CI failures on PHP 5.2. 5baea012feed2f792ab395d7ba5c5b183ae0e68d contains a fix but it looks like that was lost.

AnnaMag replied:
Reapplying.

AnnaMag commented:
We will pick it up shortly when cleaning JP Track events.

oskosk committed:
User get_option default. Was _Replace ternary with option default_

oskosk committed:
fix whitespace

oskosk committed:
Remove trailing comma

jeherve commented:
Is this ready for another review?

AnnaMag committed:
Passing blogid as an object to tracks init f-n.

It will end up in superprops. At the moment it is the only place
we set it. It will need to be specified manually unless we want
to change functions signature or pass Tracks as props to the analytics object.

AnnaMag commented:
Update on the changes:

This change affects ONLY the analytics object managing client-side tracks. If we want to keep f-n signature unchanged and chose to define event properties via object, placing extra field in superProps manually seems like a valid way to go.
As such, do not need to use `getAll()` function:  https://github.com/Automattic/jetpack/pull/9971/files#diff-6e93f8187304fd194459364df074615cL132.

AnnaMag commented:
Note: we should follow-up with respective changes to ` _inc/lib/tracks/tracks-callables.js`.

jeherve commented:
I see something different on my end right now:

![image](https://user-images.githubusercontent.com/426388/57237164-ea17b080-7026-11e9-871a-306e59942757.png)
![image](https://user-images.githubusercontent.com/426388/57237181-f7cd3600-7026-11e9-9124-ed7c073a16dd.png)

Shouldn't `eventProperties` include the blog ID?

AnnaMag commented:
@jeherve—Ah, that was a missed change. It has been corrected now, I'd appreciate another review. Thanks!

AnnaMag committed:
Remove left over from name and replace it with the correct argument.

tyxla commented on _inc/lib/tracks/class.tracks-client.php:
```diff
@@ -184,6 +184,7 @@ static function get_connected_user_tracks_identity() {
 		}
 
 		return array(
+			'blogid' => Jetpack_Options::get_option( 'id', 0 ),
```
Nit: now that we're here, should we align those `=>` vertically according to our coding standards?

tyxla commented on _inc/client/main.jsx:
```diff
@@ -118,8 +118,9 @@ class Main extends React.Component {
+
 		if ( tracksUser ) {
-			analytics.initialize( tracksUser.userid, tracksUser.username );
+			analytics.initialize( tracksUser.userid, tracksUser.username, { blogid: tracksUser.blogid } );
```
I'm a bit concerned that we're initializing analytics in several locations, but setting `blogid` only in this one. Should we make sure that any initialization of analytics throughout Jetpack can benefit from this change?

AnnaMag replied:
This is the only instance that uses that client-side library. The other with similar names are PHP-side events.

tyxla replied:
@AnnaMag I am talking about these instances, which don't seem to be addressed in this PR:

https://github.com/Automattic/jetpack/blob/4e716c7bdf6aae6dd5fe39c9029a509c7c2bee6f/_inc/idc-notice.js#L21

https://github.com/Automattic/jetpack/blob/f8078c2cd12ac508334da2fb08e37a92cf283c14/modules/widgets/search/js/search-widget-admin.js#L16

Thoughts?

AnnaMag replied:
they are handled by another tracks emitting module on the servers-side. This is not part of this PR.

tyxla reviewed (dismissed):
This seems to work well in my initial testing 👍 

However, I'm concerned that we're not adding the `blogid` in every instance of initialization of the analytics module. This will result in some events still lacking the `blogid` event prop.

AnnaMag commented:
I removed the potentially confusing line, which belongs to the server-side library and initial attempts to handle multiple changes at once ( at that time, I did not have enough understanding of tackling tracks in Jetpack ). This change and testing refers to the client-side JS code, and has atm only a single init. instance.

tyxla commented:
@AnnaMag I might be missing something, but see https://github.com/Automattic/jetpack/pull/9971#discussion_r281974315 for the other instances I am talking about.

AnnaMag commented:
@tyxla answered above.

tyxla reviewed (dismissed):
Given that we made a conscious decision not to target the other instances I was talking about, this LGTM 👍

jeherve reviewed (commented):
It's odd, I'm still not quite seeing what I should be seeing, and `this.props.tracksUserData` does not even include the blog ID on my test site, even though it is connected to WordPress.com:

![image](https://user-images.githubusercontent.com/426388/57373927-e6586b00-7199-11e9-9717-8a39fa16ec77.png)
<img width="1426" alt="screenshot 2019-05-08 at 13 58 25" src="https://user-images.githubusercontent.com/426388/57373936-ed7f7900-7199-11e9-9de8-a545e338ab98.png">

AnnaMag committed:
Use blog_id property to be picked up but Tracks ETL and set in its respetive higher level property.

jeherve reviewed (approved):
this seems to work well for me now. 👍
